[2025-06-07 Fiddler](https://thefiddler.substack.com/p/can-you-crack-the-roman-code)
====================

Fiddler
-------
This kind of problem is easier to solve with code than with thinking.

In [1]:
from functools import cache

@cache
def combinations(code, digits):
    count = 0
    for digit in digits:
        if code == digit:
            count += 1
        elif code[0:len(digit)] == digit:
            count += combinations(code[len(digit):], digits)
    return count

In [2]:
combinations("IIIIIIIIII",frozenset({"I","II","III"}))

274

Extra credit
------------

In [3]:
combinations("IIIVIIIVIIIVIII",frozenset({"I","II","III","IV","V","VI","VII","VIII"}))

4000

With thinking
-------------
To get the answers by thinking instead of code, work backwards.

|Remaining code|ways|
|--------------|----|
|I             |1   |
|II            |1+I = 2 |
|III           |1+I+II = 4 |
|IIII          |I+II+III = 7 |
|IIIII         |II+III+IIII = 13 |
|IIIIII        |III+IIII+IIIII = 24 |
|IIIIIII       |IIII+IIIII+IIIIIII = 44 |
|IIIIIIII      |IIIII+IIIIII+IIIIIII = 81 |
|IIIIIIIII     |IIIIII+IIIIIII+IIIIIIII = 149 |
|IIIIIIIIII    |IIIIIII+IIIIIIII+IIIIIIIII = 274 |

This is [A000073 in OEIS](https://oeis.org/A000073), the "tribonacci numbers".

|Remaining code |ways|
|---------------|----|
|I              |1   |
|II             |1+I = 2 |
|III            |1+I+II = 4 |
|VIII           |1+I+II+III = 8 |
|IVIII          |III+VIII = 12 |
|IIVIII         |VIII+IVIII = 20 |
|IIIVIII        |VIII+IVIII+IIVIII = 40 |
|VIIIVIII       |VIII+IVIII+IIVIII+IIIVIII = 80 |
|IVIIIVIII      |IIIVIII+VIIIVIII = 120 |
|IIVIIIVIII     |VIIIVIII+IVIIIVIII = 200 |
|IIIVIIIVIII    |VIIIVIII+IVIIIVIII+IIVIIIVIII = 400 |
|VIIIVIIIVIII   |VIIIVIII+IVIIIVIII+IIVIIIVIII+IIIVIIIVIII = 800 |
|IVIIIVIIIVIII  |IIIVIIIVIII+VIIIVIIIVIII = 1200 |
|IIVIIIVIIIVIII |VIIIVIIIVIII+IVIIIVIIIVIII = 2000 |
|IIIVIIIVIIIVIII|VIIIVIIIVIII+IVIIIVIIIVIII+IIVIIIVIIIVIII = 4000 |